bot/internal/bot/changelog.go

```diff
@@ -0,0 +1,159 @@
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bot
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+	"k8s.io/utils/strings/slices"
+)
+
+const NoChangelogLabel string = "no-changelog"
+const ChangelogPrefix string = "changelog: "
+const ChangelogTag string = "changelog"
+
+func (b *Bot) Changelog(ctx context.Context) error {
+	pull, err := b.c.GitHub.GetPullRequest(ctx,
+		b.c.Environment.Organization,
+		b.c.Environment.Repository,
+		b.c.Environment.Number,
+	)
+	if err != nil {
+		return trace.Wrap(err, "failed to retrieve pull request for https://github.com/%s/%s/pull/%d", b.c.Environment.Organization, b.c.Environment.Repository, b.c.Environment.Number)
+	}
+
+	if slices.Contains(pull.UnsafeLabels, NoChangelogLabel) {
+		log.Printf("PR contains no changelog label %q, skipping changelog check", NoChangelogLabel)
+		return nil
+	}
+
+	changelogEntry, err := b.getChangelogEntry(ctx, pull.UnsafeBody)
+	if err != nil {
+		return trace.Wrap(err, "failed to get changelog entry")
+	}
+
+	err = b.validateChangelogEntry(ctx, changelogEntry)
+	if err != nil {
+		return trace.Wrap(err, "failed to validate changelog entry")
+	}
+
+	err = b.deleteComments(ctx)
+	if err != nil {
+		return trace.Wrap(err, "failed to delete changelog comments")
+	}
+
+	return nil
+}
+
+func (b *Bot) getChangelogEntry(ctx context.Context, prBody string) (string, error) {
+	changelogRegexString := fmt.Sprintf("(?mi)^%s.*", regexp.QuoteMeta(ChangelogPrefix))
+	changelogRegex, err := regexp.Compile(changelogRegexString)
+	if err != nil {
+		return "", trace.Wrap(err, "failed to compile changelog prefix regex %q", changelogRegexString)
+	}
+
+	changelogMatches := changelogRegex.FindAllString(prBody, 2) // Two or more changelog entries is invalid, so no need to search for more than two.
+	if len(changelogMatches) == 0 {
+		return "", b.logFailedCheck(ctx, "Changelog entry not found in the PR body. Please add a %q label to the PR, or a changelog line starting with `%s` followed by the changelog entry for the PR.", NoChangelogLabel, ChangelogPrefix)
+	}
+
+	if len(changelogMatches) > 1 {
+		return "", b.logFailedCheck(ctx, "Found two or more changelog entries in the PR body: %v", changelogMatches)
+	}
+
+	changelogEntry := changelogMatches[0][len(ChangelogPrefix):] // Case insensitive prefix removal
+	log.Printf("Found changelog entry %q matching %q", changelogEntry, changelogRegexString)
```

Comment:
Nit: Don't think we really need to log the regex.
```suggestion
	log.Printf("Found changelog entry: %q", changelogEntry)
```

Reply:
I somewhat disagree here - I have found regex logging to be helpful in the past, especially when testing an update to regex logic. That being said, if the regex is hard-coded as discussed above then you are right - this log entry is probably unnecessary.